### PR TITLE
Immutable filtering and searching methods

### PR DESCRIPTION
**Pull request by mxstbr**
Refactored filtering and searching methods to make query immutable.

This is much cleaner than what we previously had, where `addFilterToQuery` and `addSearchToQuery` would mutate the query somewhere deep inside the fields.

### CODE REVIEW

File: admin/server/api/list/download.js [hunk 1]
```diff
@@ -11,10 +12,10 @@ module.exports = function (req, res) {
 		catch (e) { } // eslint-disable-line no-empty
 	}
 	if (typeof filters === 'object') {
-		req.list.addFiltersToQuery(filters, where);
+		where = assign({}, req.list.addFiltersToQuery(filters));
 	}
 	if (req.query.search) {
-		req.list.addSearchToQuery(req.query.search, where);
+		where = assign({}, req.list.addSearchToQuery(req.query.search));
```
Comment (JedWatson) [inline admin/server/api/list/download.js]: It's a valid use case to provide _both filters and a search_, so these lines should be

```
assign(where, req.list.addSearchToQuery(req.query.search));
```

Comment (mxstbr) [inline admin/server/api/list/download.js]: Oh yeah, good catch, my bad. Will fix!

**Comment by mxstbr** [comment]
Fixed and ready to merge!

**Comment by JedWatson** [comment]
Looks good!

**Pull request merged**